externals/llvm-external-projects/torch-mlir-dialects/include/torch-mlir-dialects/Dialect/Tcp/IR/TcpOps.td

```diff
@@ -0,0 +1,65 @@
+#ifndef TORCH_MLIR_DIALECT_TCP_OPS
+#define TORCH_MLIR_DIALECT_TCP_OPS
+
+include "torch-mlir-dialects/Dialect/Tcp/IR/TcpOpsBase.td"
+
+include "mlir/IR/OpBase.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+def Tcp_TanhOp : Tcp_UnaryOp<"tanh", [NoSideEffect]> {
+  let summary = "Computes tanh of input, elementwise";
+
+  let description = [{
+    Elementwise unary op to compute tanh.
+  }];
+
+  let arguments = (ins
+    Tcp_Tensor:$in
+  );
+
+  let results = (outs
+    Tcp_Tensor:$out
+  );
+
+  let assemblyFormat = "$in attr-dict `:` type($in) `->` type($out)";
+}
+
+def Tcp_AddOp : Tcp_BinaryOp<"add", [NoSideEffect]> {
+  let summary = "Computes elementwise addition";
+
+  let description = [{
+    Elementwise addition.
+  }];
+
+  let arguments = (ins
+    Tcp_Tensor:$in1,
+    Tcp_Tensor:$in2
+  );
+
+  let results = (outs
+    Tcp_Tensor:$out
+  );
+
+  let assemblyFormat = "$in1 `,` $in2 attr-dict `:` type($in1) `,` type($in2) `->` type($out)";
+}
+
+def Tcp_MatmulOp : Tcp_Op<"matmul", [NoSideEffect]> {
+  let summary = "Computes 2D / 3D matrix multiplication";
+
+  let description = [{
+    Matrix multiplication of two 2D / 3D tensors.
```

Comment:
This is a good place to define the semantics of this op in more detail (enough detail for someone to implement a reference interpreter).

Matmul semantics are probably obvious, but this op will set the tone for other less obvious ops that come after it.

Reply:
It is honestly unclear to me if we want to support `(2D,3D)` mixed cases or require both to be the same rank. Or something more like dot_general

Reply:
IMO because something more general like `torch.tensordot` or `tf.tensordot` can be mapped to `tcp.matmul` with sequence of operand transposes & reshapes, and the bottom consumers transform for backends (codegen or lib-calls) generally prefers lower ranks matmuls/batch_matmul contracts `e.g match(AB), match(A^TB), match(AB^T)..etc`. In this case given that how TCP is positioned in the stack a lower rank fixed contract `reduce_add(A[i,m,k] * B[i,k,n], k)` make sense.

Reply:
Good point regarding the mixed 2D-3D case. In the form I had, that case is not allowed and we will have to error out.